sys/net/application_layer/coapfileserver/coapfileserver.c

```diff
@@ -0,0 +1,407 @@
+        char *entry_name = entry.d_name;
+        size_t entry_len = strlen(entry_name);
+        if (entry_len <= 2 && memcmp(entry_name, "..", entry_len) == 0) {
+            /* Up pointers don't work the same way in URI semantics */
+            continue;
+        }
+        bool is_dir = entry_is_dir(request->namebuf, entry_name);
+        /* maybe ",", "<>", and the length */
+        ssize_t need_bytes = (payload_cursor == 0 ? 0 : 1) + 2 + entry_len + (is_dir ? 1 : 0);
+        if (payload_cursor + need_bytes > pdu->payload_len) {
+            /* Without blockwise, this is the best approximation we can do */
+            DEBUG("coapfileserver: Directory listing truncated\n");
+            break;
+        }
+        if (payload_cursor != 0) {
+            pdu->payload[payload_cursor++] = ',';
+        }
+        pdu->payload[payload_cursor++] = '<';
+
+        payload_cursor += _copy_string(&pdu->payload[payload_cursor], resource->resource,
+                                       pdu->payload_len - payload_cursor);
+        payload_cursor += _copy_string(&pdu->payload[payload_cursor], &request->namebuf[root_len],
+                                       pdu->payload_len - payload_cursor);
+        pdu->payload[payload_cursor++] = '/';
+        payload_cursor += _copy_string(&pdu->payload[payload_cursor], entry_name,
+                                       pdu->payload_len - payload_cursor);
+        if (is_dir) {
+            pdu->payload[payload_cursor++] = '/';
+        }
+        pdu->payload[payload_cursor++] = '>';
+    }
+    vfs_closedir(&dir);
+
+    if (payload_cursor == 0) {
+        /* Rewind to clear payload marker */
+        payload_cursor -= 1;
+    }
+
+    return resp_len + payload_cursor;
+}
+
+static size_t coapfileserver_errno_handler(coap_pkt_t *pdu, uint8_t *buf, size_t len, int err)
+{
+    uint8_t code;
+    switch (err) {
+        case -EACCES:
+            code = COAP_CODE_FORBIDDEN;
+            break;
+        case -ENOENT:
+            code = COAP_CODE_PATH_NOT_FOUND;
+            break;
+        default:
+            code = COAP_CODE_INTERNAL_SERVER_ERROR;
+    };
+    DEBUG("coapfileserver: Rejecting error %d (%s) as %d.%02d\n", err, strerror(err),
+          code >> 5, code & 0x1f);
+    return gcoap_response(pdu, buf, len, code);
+}
+
+static int stat_etag(char *filename, union stattag *stattag)
+{
+    int err = vfs_stat(filename, &stattag->stat);
+    if (err < 0) {
+        return err;
+    }
+
+    /* Normalizing fields whose value can change without affecting the ETag */
+    stattag->stat.st_nlink = 0;
+    memset(&stattag->stat.st_atime, 0, sizeof(stattag->stat.st_atime));
+
+    /* Build a compact ETag by XORing the stat onto itself */
+    for (unsigned int i = 0; i < sizeof(stattag->etag_padding) / sizeof(stattag->etag); ++i) {
+        for (int j = 0; j < ETAG_LENGTH; ++j) {
+            stattag->etag[j] ^= stattag->etag_padding[i * ETAG_LENGTH + j];
+        }
+    }
```

Comment:
Why not use a checksum algorithm here?

Reply:
Pure laziness on my part -- a checksum algorithm would be perfectly suitable.

(aiocoap-fileserver even uses a truncated SHA to build an ETag because that's what's easy to use there. I just XOR'd because that was quicker to write than looking up which checksum algorithm would likely already be part of RIOT).